airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py

```diff
@@ -0,0 +1,252 @@
+
+    Therefore, we need to manage state per partition.
+    """
+
+    _NO_STATE = {}
+    _NO_CURSOR_STATE = {}
+    _KEY = 0
+    _VALUE = 1
+
+    def __init__(self, cursor_factory, partition_router):
+        self._cursor_factory = cursor_factory
+        self._partition_router = partition_router
+        self._cursor_per_partition = {}
+
+    def stream_slices(self, sync_mode: SyncMode, stream_state: StreamState) -> Iterable[PerPartitionStreamSlice]:
+        """
+        We knowingly avoid using stream_state as we want PerPartitionCursor to manage its own state.
+        """
+        slices = self._partition_router.stream_slices(sync_mode, self._NO_STATE)
+        for partition in slices:
+            cursor = self._cursor_per_partition.get(self._to_partition_key(partition))
+            if not cursor:
+                cursor = self._create_cursor(self._NO_CURSOR_STATE)
+                self._cursor_per_partition[self._to_partition_key(partition)] = cursor
+
+            for cursor_slice in cursor.stream_slices(sync_mode, self._get_state_for_partition(partition)):
+                yield PerPartitionStreamSlice(partition, cursor_slice)
+
+    def update_cursor(self, stream_slice: PerPartitionStreamSlice, last_record: Optional[Record] = None):
+        if not last_record:
+            # The `update_cursor` method is called without `last_record` in order to set the initial state. In that case, stream_slice is
+            # not a PerPartitionStreamSlice but is a dict representing the state
+            self._init_state(stream_slice)
+        else:
+            try:
+                self._cursor_per_partition[self._to_partition_key(stream_slice.partition)].update_cursor(
+                    stream_slice.cursor_slice, last_record
+                )
+            except KeyError as exception:
+                raise KeyError(
+                    f"Partition {str(exception)} could not be found in current state based on the record. This is unexpected because "
+                    f"we should only update cursor for partition that where emitted during `stream_slices`"
+                )
+
+    def _init_state(self, stream_state: dict) -> None:
+        if not stream_state:
+            return
+
+        for state in stream_state["states"]:
+            self._cursor_per_partition[self._to_partition_key(state["partition"])] = self._create_cursor(state["cursor"])
+
+    def get_stream_state(self) -> StreamState:
+        states = []
+        for partition_tuple, cursor in self._cursor_per_partition.items():
+            cursor_state = cursor.get_stream_state()
+            if cursor_state:
+                states.append(
+                    {
+                        "partition": self._to_dict(partition_tuple),
+                        "cursor": cursor.get_stream_state(),
+                    }
+                )
+        return {"states": states}
+
+    def _get_state_for_partition(self, partition: Mapping[str, Any]) -> Any:
```

Comment:
nit: doesn't this always return a `StreamState`?

Reply:
`Optional[StreamState]`, yes. Updated!